tests/src/python/test_qgsprojectbadlayers.py

```diff
@@ -219,8 +219,9 @@ def _check_relations():
             _check_relations()
 
         # Changing data source, relations should be restored:
-        point_a.setDataSource(point_a_source, 'point_a', 'ogr')
-        point_b.setDataSource(point_b_source, 'point_b', 'ogr')
+        options = QgsDataProvider.ProviderOptions()
+        point_a.setDataSource(point_a_source, 'point_a', 'ogr', options)
```

Comment:
Can we add this as a new test?
The old way should continue to work.

Reply:
The old way does not works, it is based on an unvirtual deprecated method only defined for QgsVectorLayer 88c0d3f07db541b1cb6d5cc81f4aabc12741de26

I can move this method to QgsMapLayer to keep API

Reply:
Can't we keep it where it was?

Reply:
If I keep it where it was, the compilation failed because this method is marked as deprecated.

Reply:
This can normally be avoided by `Q_NOWARN_DEPRECATED_PUSH` / `Q_NOWARN_DEPRECATED_POP`
Removing it is an API break, so we can't do that.
Moving it to QgsMapLayer is not nice because it increases exposure of a deprecated API, ideally we should avoid that.

Reply:
It seems that keeping the dedicated QgsVectorLayer::setDataSource method in place. The compilation failed.

I need to explicitly call the QgsMapLayer::setDataSource method

Reply:
It seems that keeping the dedicated QgsVectorLayer::setDataSource method in place. It is not possible to use QgsMapLayer::setDataSource in PyQGIS from QgsVectorlayer instance.